src/vc_util/src/lib.rs

```diff
@@ -0,0 +1,421 @@
+use candid::Principal;
+use canister_sig_util::{extract_raw_canister_sig_pk_from_der, CanisterSigPublicKey};
+use ic_certified_map::Hash;
+use ic_crypto_standalone_sig_verifier::verify_canister_sig;
+use ic_types::crypto::threshold_sig::IcRootOfTrust;
+use identity_core::convert::FromJson;
+use identity_credential::credential::Subject;
+use identity_credential::error::Error as JwtVcError;
+use identity_credential::validator::JwtValidationError;
+use identity_jose::jwk::{Jwk, JwkParams, JwkParamsOct, JwkType};
+use identity_jose::jws::{
+    CompactJwsEncoder, Decoder, JwsAlgorithm, JwsHeader, SignatureVerificationError,
+    SignatureVerificationErrorKind,
+};
+use identity_jose::jwt::JwtClaims;
+use identity_jose::jwu::{decode_b64, encode_b64};
+use serde_json::Value;
+use sha2::{Digest, Sha256};
+use std::ops::{Add, Deref, DerefMut};
+
+pub const II_CREDENTIAL_URL_PREFIX: &str =
+    "https://internetcomputer.org/credential/internet-identity/";
+pub const II_ISSUER_URL: &str = "https://internetcomputer.org/issuers/internet-identity";
+pub const VC_SIGNING_INPUT_DOMAIN: &[u8; 26] = b"iccs_verifiable_credential";
+
+/// A pair of identities, that denote the same user.
+/// Used in attribute sharing flow to maintain II's unlinkability of identities.
+pub struct AliasTuple {
+    /// A temporary identity, used in attribute sharing flow.
+    pub id_alias: Principal,
+    /// An identity under which a user is known to a dapp.
+    pub id_dapp: Principal,
+}
+
+#[derive(Debug)]
+pub enum CredentialVerificationError {
+    InvalidJws(SignatureVerificationError),
+    InvalidClaims(JwtValidationError),
+}
+
+/// Returns the effective bytes that will be signed when computing a canister signature for
+/// the given JWT-credential, verfiable via the specified public key.
+pub fn vc_signing_input(
+    credential_jwt: &str,
+    canister_sig_pk: &CanisterSigPublicKey,
+) -> Result<Vec<u8>, String> {
+    let encoder = jws_encoder(credential_jwt, canister_sig_pk)?;
+    Ok(encoder.signing_input().to_vec())
+}
+
+/// Computes and returns SHA-256 hash of the given `signing_input` prefixed with
+///      length(VC_SIGNING_INPUT_DOMAIN) || VC_SIGNING_INPUT_DOMAIN
+/// (for domain separation).
+pub fn vc_signing_input_hash(signing_input: &[u8]) -> Hash {
+    let mut hasher = Sha256::new();
+    let buf = [VC_SIGNING_INPUT_DOMAIN.len() as u8];
+    hasher.update(buf);
+    hasher.update(VC_SIGNING_INPUT_DOMAIN);
+    hasher.update(signing_input);
+    hasher.finalize().into()
+}
+
+/// Constructs and returns a JWS (a signed JWT) from the given components.
+pub fn vc_jwt_to_jws(
+    credential_jwt: &str,
+    canister_sig_pk: &CanisterSigPublicKey,
+    sig: &[u8],
+) -> Result<String, String> {
```

Comment:
pushed to later -- these errors should never happen (as they would indicate some internal corruption), and I had panics on them before, but switched to `Result` to avoid panics.